# APS106 Lecture Notes - Week 1, Lecture 3

### This Week
| Lecture | Topics | Reading |
| --- | --- | --- | 
| 1.1 | Introduction | |
| 1.2 | The Coding Toolbox | [Chapter 1](https://learn.zybooks.com/zybook/UTORONTOAPS106Winter2024/chapter/1/section/1)  |
| **1.3** | **Variables, Expressions, and Operators** | [Chapter 2](https://learn.zybooks.com/zybook/UTORONTOAPS106Winter2024/chapter/2/section/1) |


### Lecture Structure
1. [Using Python as a Calculator](#section1)
2. [Variables and Memory](#section2)
3. [Different Types of Variables](#section3)
4. [Augmented Assignment Operators](#section4)
5. [Let's Code!](#section5)

## Recap: types of cells in a Jupyter notebook

Text, image, and code!

## Recap: What is Programming?

- A way of telling a computer what to do.
- A computer can’t infer (…yet).
    - Need to tell a computer every single step in a language it can understand.

How would you request an egg for breakfast to a chef and to a computer/robot?

|<span style="color:darkgreen"> **Chef** </span> | <span style="color:darkred"> **Computer** </span>
| --- | --- |
| Sunny-side up, please! | Turn on stove
|| Take out pan
|| Take one egg out of fridge
|| Crack egg
|| Pour egg into pan
|| Wait 5 minutes 

<a id='section1'></a>

## 1. Using Python as a Calculator

Unlike some other languages, Python is dynamically typed.  It is interpreted language and can run line by line.  

### Arithmetic Operators

| Operator | Operation | Expression | English | Result |
| --- | --- | --- | --- | --- |
| + | addition | 11 + 23 | 11 plus 23 | 34 |
| - | subtraction | 23 - 52 | 23 minus 52 | -29 |
| * | multiplication | 4 * 5 | 4 times 5 | 20 |
| / | division | 9 / 2 | 9 divided by 2 | 4.5 |
| // | integer division | 9 // 2 | 9 divided by 2 | 4 |
| ** | exponentiation | 2 ** 5 | 2 to the power 5 | 32 |
| % | modulo (remainder) | 9 % 2 | 9 mod 2 | 1 |


The first 3 (addition, subtraction, multiplication) should be easy - with the slight issue that multiplication is represented by `*`. This is pretty common in (almost) all programming languages.

#### Division

Division comes in two flavours.

First, the `/` operator divides a number by another number (integer or decimal) and returns and represents the result **as a `float`** (a value type with decimal places - we'll discuss later this lecture).

```
10/3 = 3.333333333333333
```

The second division operator is integer division. It divides the first number by the second with the result being **truncated** to an integer value. "Truncated" means that everything after the decimal is thrown away.  NOTE THAT WE SAID TRUNCATED, NOT ROUNDED.

```
10/3 = 3.0
```

If the two operands of `//` (the two numbers being divided) are both `int`, then the result will be an `int`. If either of the two operands is a `float`, then the result will be a `float` (even though it will have a whole number value).

#### Exponentiation

Python has a nice exponentiation operator `**`

#### Modulo

The last operator may be new to you. It is the modulo operator repesented by `%`. It is also a kind of division but it results in the **remainder** after dividing the operands. 

In [1]:
3+4

7

In [2]:
4*5

20

In [3]:
10/3

3.3333333333333335

In [4]:
10/2

5.0

In [5]:
10//3

3

In [6]:
2**3

8

In [7]:
10%3

1

### Arithmetic Operator Precedence

Just as you learned in high school, when you have a complicated expression some of the operators are executed before others. In Python the operator precedence is as follows.

| Operator | Precedence |
| --- | --- | 
| ** | highest | 
| - (negation) | | 
| \*, /, //, % | (evaluate left-to-right) | 
| + (addition, - (subtraction) | lowest (evaluate left-to-right) |
See Gries, Table 2, p. 15.

In [8]:
25 + 2 ** 2

29

What would be the output of the following expression?
```
100 - 25 * 2 ** 2
```

In [14]:
(100 -25 )* 2 ** 2

300

In [9]:
2 ** 2

4

In [10]:
25 * 4

100

<a id='section2'></a>

## 2. Variables and Memory

The most basic thing you can do in a computer program is to assign a value to a variable. We'll get into more detail later in this lecture, but for now, you can think of a variable as a **named** location in memory and the value as the bits (i.e., 0/1 values) that represents the quantity that is stored in that location.

Here are three lines in a Python program. Each line does 3 things:
- creates a variable
- gives it a name
- assigns a value

In [16]:
x = 2

In [17]:
x

2

In [18]:
base = 20
height = 12
area = base*height

In [19]:
area

240

So now we have a variable named `base` with a value of 20 and a variable named `height` with a value of 12.  There is also a variable named area with a value of 20 x 12 (which equals 240).  But we didn't get any output?!

### Output: The `print` statement

You will have noticed when I ran the above code, it didn't look like anything happened. Stuff did happen (as described above) but since there was no output, you couldn't tell. 

We'll get into this in (much more) depth later in the course. For now, you should know that there exists a function called `print` which prints a value to the screen. Whenever you need to output something, you will need to use a `print` function.

For example:

In [20]:
print(20)

20


In [21]:
print(base)

20


In [22]:
print(height)

12


In [23]:
print(base*height)
print(area)

240
240


### Variable Names

Python has a two rules for legal variable names:
1. Names must start with a letter or \_ (an underscore).
2. Names must contain only letters, digits, and \_. [No emojis in your variables names!]

By convention, it is good Python style to use **`snake_case`**. Create meaningful variables names with underscore separating the "words" in the name (if there is more than one meaningful word).

Object `family_name = "moosavi"` not found.


family_name





In [ ]:



time_to_launch = 20 # maybe used in some sort of aerospace application?

family_name = "Smith" # for a human resources application

learning_rate = 0.01 # for machine learning code

#HINT: THE FOLLOWING ARE NOT HELPFUL NAMES
nom = 1.2
nomnomnom = 0.0000012
nomnomnomnomnom = 5
nomnomnomnomnomnomnom = 'nomnom string'

### Memory Locations

*Everything* in Python has a location in memory. **This is different than most languages!** Each location in memory has an *address*. You can think of this as similar to a house address. Just as "40 St. George St" refers to a location in Toronto, each location has a corresponding "id" which maps to its location in memory.

For example:

In [31]:
base = 20
base2 = 20

In [32]:
print(base)
print(base2)

20
20


In [34]:
print(id(base))
print(id(base2))
print(id(20))

4337919800
4337919800
4337919800


In [35]:
print(base, base2)

20 20


In [30]:
print(id(20))

4337919800


## Assignment Statements 

<div class="alert alert-block alert-warning">
This is a frequent source of error and mis-understanding.
</div>

An assignment statement is always executed as follows:
1. The expression to the right of the `=` sign is evaluated. "Evaluated" means that the computer figures out what quantity it represents. That quantity is someplace in memory (i.e., it has a memory address) and that address is figured out.
2. The memory address of the evaluated expression is then stored in the variable on the left side of the `=` sign.

The notion of evaluation may sound strange at the moment because a numeric constant (like 20) just evaluates to itself. So it seems like this is more complicated than it has to be. As we will see, we can have more than just a constant on the right-hand side of an assignment. And memorizing the following rule will help you in those more complicated cases: 

<font color=red>first evaluate the thing on the right, then assign its address to the thing on the left</font>.

Let's go through a simple example.

In [36]:
difference = 20
print(difference)

20


Following our rule, the thing in the right of the `=` is evaluated (to 20, obviously) and then it's address is figured out. The address of 20 is:

In [37]:
print(id(difference))
print(id(20))


4337919800
4337919800


In [38]:
double = 2 * difference
print(double)

40


In [39]:
print(id(double))
print(id(40))

4337920440
4337920440


In [40]:
difference = 5
print(difference)

5


In [41]:
print(id(difference))
print(id(5))
print(id(20))

4337919320
4337919320
4337919800


<a id='section3'></a>
<div class="alert alert-block alert-info">
<big><b>Where Are We So Far</b></big>
<ul>  
    <li><tt>print</tt> statements to output something to the screen</li>  
 <li>assignment statements that create a variable and assign it a value</li>  
 <li>the rule for figuring out what assignment statements are really doing</li>  
</ul>  
</div>

### 3. Different Types of Variables

So far we've mostly seen examples of assigning numbers to variables. And mostly, we've only seen a sub-class of numbers being assigned: integers.

To start, let's talk about 3 types of variables: `int` (represents integers), `float` (represents continuous numbers), and `str` (represents a strong of one or more characters).

A *type* is a set of values and a set of operations (e.g., +,-, etc. we'll see more of these in a second) that can be performed on the values. 

#### Types `int` and `float`

Numeric values are represented in Python as `int` and `float` type. (`float` is short for "floating-point" number which has to do with how you represent fractional numbers with only bits. Not something we have to be concerned about for now.) We've seen examples of `int` variables, here are a couple of `float` variables.

In [43]:
integer_variable = 1
print(type(integer_variable))

<class 'int'>


In [44]:
string_variable = "APS106"
print(type(string_variable))

<class 'str'>


In [42]:
distance = 5.3
print(distance)
time = 13.4
print(time)
print(type(time))

5.3
13.4
<class 'float'>


In [ ]:
approx_time = 13
print(type(approx_time))

#### Type `str`

A *string literal* is a sequence of characters. In Python, this type is called `str`. Strings in Python start and end with a single quotes (') or double quotes (") (or even triple quotes - but we'll get to that later). Just remember that you need to start and end a string literal with the same quote mark. A string can be made up of letters, numbers, and special characters. For example:

In [46]:
print("hello\nagain") #we will talk more about what "\n" does in the future, can you guess?

hello
again


In [47]:
greeting = 'hello'
print(greeting)

print('how are you?')
print("short- and long-term")

hello
how are you?
short- and long-term


If you start a string literal with a single-quote you need to end it with a single-quote. Similarly, if it starts with a double-quote, a double-quote ends it.

In [48]:
welcome = "Welcome to 'APS106'!"
print(welcome)

Welcome to 'APS106'!


In [49]:
bad_str = 'Start and end"
print(bad_str)

SyntaxError: unterminated string literal (detected at line 1) (3951288836.py, line 1)

Just like an `int` and `float` variables store integers and floating point values, a `str` variable stores string literals.

### Expressions

You already know about mathematical expressions like `2+5-13`. Such expressions: evaluate to some value and are made up of *operators* (e.g., +, -) and *operands* (the values that the operators operate on).

Actually, expressions do not have to have operators. A number by itself or a variable is an expression. When we talked about evaluating the thing after the = sign in an assignment statement, that thing is an expression. So this means we can expand our assignment statements to have any expression on the right-hand side.

In [50]:
base = 20
height = 12
area = base * height / 2
print(area)

120.0


In [51]:
celsius = 22
fahrenheit = celsius * 9/5 + 32
print(fahrenheit)

71.6


Notice that the examples above have a variable to the right of the = sign. This is fine as a variable will just evaluate to the value that it has been assigned.

The rules for figuring out what assignments statements are doing remain the same. The change is that the evaluation of the expression on the right-hand side is more complicated when you have something other than a number. 

<a id='section4'></a>
## 4. Augmented Assignment Operators


| Operator | Expression | Identical Expression | 
| --- | --- | --- | 
| += | x += 2 | x = x + 2 |
| -= | x -= 3 | x = x - 3 |
| \*= | x \*= 5 | x = x * 5 |
| /= | x /= 2 | x = x / 2 | 
| //= | x //= 2 | x = x // 2 |
| \*\*= | x \*\*= 3 | x = x ** 3 |
| %= | x %= 3 | x = x % 3 |
See Gries, Table 3, p. 22.

The thing to the right of the augmented operator is just an expression and so you can do more complicated things.

It is often the case that we want to evaluate some expression using a variable and store that calculation back in the same variable. For example, if you wanted to count something (like the number of perfect squares less than or equal to some number that a user enters), it makes sense to add one to a variable every time you find another perfect square.

Let's write that code.

In [ ]:
max_num = int(input("Give me an integer: "))

base = 1
count = 0
while base * base <= max_num:
    count = count + 1
    base = base + 1

print("There are", count, "perfect squares less than", max_num)


We can re-write the above code using "augmented operators" that combine `+` and `=`.

In [ ]:
max_num = int(input("Give me an integer: "))

base = 1
count = 0
while base * base <= max_num:
    count += 1
    base += 1

print("There are", count, "perfect squares less than", max_num)

This code does exactly the same thing: we've just abbreviated by using augmented operators.

We can do this with all the standard arithmetic operators.

In [ ]:
x = 17
x //= 3
print(x)

In [ ]:
y = 7
y *= 2
print(y)

In [ ]:
z = 20
z %= 6
print(z)

In [ ]:
#remember: evaluate right hand side of equation first

x = 7
y = 3
x += y**2 - (4 * 3) - y
print(x)

The standard rules apply: evaluate the expression on the right-hand side and then increment (in the case of `+=`) the variable on the left-hand side.

<div class="alert alert-block alert-info">
<big><b>Where Are We Now</b></big>
<ul>  
    <li>Three variables types: <tt>int</tt>, <tt>float</tt>, and <tt>str</tt></li>
    <li>Arithmetic expression combine operators, variables, constants</li>
    <li>Arithmetic operators have precedence which can be over-ridden with parentheses</li>
        <li>Augmented operators combine an operator with an assignment back to the variable being operated on</li>

</ul>  
</div>


# List of Python Operators

### 1. Arithmetic Operators
- `+` : Addition
- `-` : Subtraction
- `*` : Multiplication
- `/` : Division
- `%` : Modulus
- `**` : Exponentiation
- `//` : Floor Division

### 2. Assignment Operators
- `=` : Assign
- `+=` : Add and assign
- `-=` : Subtract and assign
- `*=` : Multiply and assign
- `/=` : Divide and assign
- `%=` : Modulus and assign
- `**=` : Exponent and assign
- `//=` : Floor division and assign
---
We learned up to here today, we will learn more operators in the following lectures...

---

- `&=` : Bitwise AND and assign
- `|=` : Bitwise OR and assign
- `^=` : Bitwise XOR and assign
- `>>=` : Right shift and assign
- `<<=` : Left shift and assign

### 3. Comparison Operators
- `==` : Equal to
- `!=` : Not equal to
- `>` : Greater than
- `<` : Less than
- `>=` : Greater than or equal to
- `<=` : Less than or equal to

### 4. Logical Operators
- `and` : Logical AND
- `or` : Logical OR
- `not` : Logical NOT

### 5. Bitwise Operators
- `&` : Bitwise AND
- `|` : Bitwise OR
- `^` : Bitwise XOR
- `~` : Bitwise NOT
- `<<` : Left shift
- `>>` : Right shift

### 6. Identity Operators
- `is` : Identity (same object)
- `is not` : Not the same object

### 7. Membership Operators
- `in` : Membership (part of)
- `not in` : Not a member


# Tips for pros (#cleancode)

</br>

- ### Use Comments:
Comments should explain why something is done, not what is done; the code itself should be self-explanatory for the latter.


- ### Use Meaningful Names: 
Choose variable, function, and class names that are descriptive and indicate what they represent or do. Avoid using vague names like data or x.

- ### Use Consistent Naming Schema
Be consistent with your naming convention, e.g., use **snake_case**. In **snake_case**, words are all in lowercase and are separated by underscores. It is a naming convention where each word is separated by an underscore character, and all letters are lowercase.


- ### Use Spaces Around Operators:
For readability, put spaces around operators and after commas: 

e.g.,
```
canda = cat + panda
```
not
```
canda=cat+panda
```
<img src="Images/canda.png" alt="canda" width="200"/>


<a id='section5'></a>
## 5. Let's Code!

In the old days (and in the United States), the mileage of a gas-powered car was measured in miles per gallon. Now for places that use the metric system, we prefer to measure it in litres per hundred kilometres. Write code to do the conversion to metric given a value in miles per gallon.

Hints:
- You don't know yet how to get the user to type in a value so just create a variable that is assigned to your miles per gallon number.
- There are 3.78541 litres in a gallon.
- There are 1.60934 kilometres in a mile.


### Step 0: Plan your code
Figure out a solution for your problem. Then convert this into a step-by-step psedo code. 

### Step 1: work out the math

Figure out on paper how to do the conversion. You could pile everything into one equation or you could take it step-by-step.

My steps:
1. calculate kilometers per gallon
1. calculate kilometers per litre
1. calculate liters per kilometer and multiple by 100

This is what we are going to call an **Algorithm Plan** in this course: the steps your code is supposed to take.

(You could also do the math a bit better and fiugure out that 1 mile-per-gallon = 235.21 litres/100k. But let's stick with the step-by-step approach for now.)

### Step 2: create your variables

In [ ]:
mile_per_gallon = 25 #this is what we want to convert - a 2019 Ford Mustang gets ~25 MPG
litre_per_gallon = 3.78541
k_per_mile = 1.60934


### Step 3: calculate kilometers per gallon

In [ ]:
k_per_gallon = mile_per_gallon * k_per_mile #mile_per_gallon & k_per_mile are given
print(k_per_gallon)

### Step 4: calculate kilometres per litre

In [ ]:
k_per_litre = k_per_gallon / litre_per_gallon #litre_per_gallon is given
print(k_per_litre)

### Step 5: calculate litres per kilometre and multiply by 100

In [ ]:
litre_per_100k = 1/k_per_litre * 100
print(litre_per_100k)

In [ ]:
#OR BY UNIT ANALYSIS:

#Miles / Gallon -> L / 100 km
# L/km = L/gal * gal/mile * mile/km

# finally, to convert L/km into L/100km, multiply by 100

Steps 1 - 5 are what we are going to call the **Programming Plan** in the course. It is the thing that you (as a programmer) have to do to create the answer to your problem.

<div class="alert alert-block alert-info">
<big><b>This Lecture</b></big>
<ul>  
    <li>Variables and Memory</li>
    <li>Variable Types</li>
    <li>Expressions and Evaluating Expressions</li>
    <li>Assignment and Augmented Assignment</li>
     <li>Arithmetic Operators</li>
         <li>Algorithm Plans and Programming Plans</li>
</ul>  
</div>
